oxidation/libparsec/crates/libparsec_fs/src/file_operations.rs

```diff
@@ -0,0 +1,358 @@
+            new_chunk.start = max(chunk.start, start);
+            new_chunk.stop = min(
+                chunk.stop,
+                NonZeroU64::new(stop)
+                    .expect("The stop offset can only be 0 if the index range is empty"),
+            );
+            new_chunk
+        })
+}
+
+pub fn prepare_read(manifest: &LocalFileManifest, size: u64, offset: u64) -> Vec<Chunk> {
+    // Sanitize size and offset to fit the manifest
+    let offset = min(offset, manifest.size);
+    let size = min(
+        size,
+        manifest
+            .size
+            .checked_sub(offset)
+            .expect("The offset computed above cannot be greater than the manifest size"),
+    );
+
+    // Find proper block indexes
+    let blocksize = u64::from(manifest.blocksize);
+    let start_block = offset / blocksize;
+    let stop_block = (offset + size + blocksize - 1) / blocksize;
+
+    // Loop over blocks
+    (start_block..stop_block)
+        .flat_map(move |block| {
+            // Get substart, substop and subsize
+            let blockstart = block * blocksize;
+            let substart = max(offset, blockstart);
+            let substop = min(offset + size, blockstart + blocksize);
+            let subsize = substop
+                .checked_sub(substart)
+                .expect("Substop is always greater than substart");
+            // Get the corresponding chunks
+            let block_chunks = manifest
+                .get_chunks(block as usize)
+                .expect("A valid manifest must have enough blocks to cover its full range.");
+            block_read(block_chunks, subsize, substart)
+        })
+        // Collect as a flatten vec of Chunks
+        .collect()
+}
+
+// Prepare write
+
+fn block_write(
+    chunks: &[Chunk],
+    size: u64,
+    start: u64,
+    new_chunk: Chunk,
+) -> (Vec<Chunk>, HashSet<ChunkID>) {
+    let stop = start + size;
+
+    // Edge case
+    if chunks.is_empty() {
+        return (vec![new_chunk], HashSet::new());
+    }
+
+    // Bisect
+    let start_index = match chunks.binary_search_by_key(&start, |x| x.start) {
+        Ok(x) => x,
+        Err(x) => x
+            .checked_sub(1)
+            .expect("First chunk should always start at 0"),
+    };
+    let stop_index = match chunks.binary_search_by_key(&stop, |x| x.start) {
+        Ok(x) => x,
+        Err(x) => x,
+    };
+
+    // Removed ids
+    let mut removed_ids: HashSet<ChunkID> = chunks
+        .get(start_index..stop_index)
+        .unwrap_or_default()
+        .iter()
+        .map(|x| x.id)
+        .collect();
+
+    // New chunks
+    let mut new_chunks: Vec<Chunk> = chunks.get(0..start_index).unwrap_or_default().to_vec();
+
+    // Test start chunk
+    let start_chunk = chunks
+        .get(start_index)
+        .expect("Indexes are found using binary search and hence always valid");
+    if start_chunk.start < start {
+        let mut new_start_chunk = start_chunk.clone();
```

Comment:
Why clone ? start_chunk not used

Reply:
The start chunk is already part of the original manifest, so we need to clone if we want to return it as part of the new one right?

Reply:
damn ref :man_facepalming: